### ISSUE TITLE
.NET Core 2.0 Preview 1: XmlSchemaSet Compile Failure to Find Types in Included Schemas

### ISSUE THREAD

**Issue by JeffFerguson** [dotnet/corefx#20263]
Consider the following XML schema:

```
<?xml version="1.0" encoding="utf-8"?>
<schema targetNamespace="http://example.com/xbrl/taxonomy" xmlns="http://www.w3.org/2001/XMLSchema" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:xbrli="http://www.xbrl.org/2003/instance" xmlns:link="http://www.xbrl.org/2003/linkbase" xmlns:xlink="http://www.w3.org/1999/xlink">
  <import namespace="http://www.xbrl.org/2003/instance" schemaLocation="http://www.xbrl.org/2003/xbrl-instance-2003-12-31.xsd"/>
  <element name="changeInRetainedEarnings" type="xbrli:monetaryItemType" substitutionGroup="xbrli:item" id="a1" xbrli:periodType="duration"/>
  <element name="fixedAssets" type="xbrli:monetaryItemType" substitutionGroup="xbrli:item" id="a2" xbrli:balance="debit" xbrli:periodType="instant"/>
</schema>
```

The following .NET Core 2.0 Preview 1 console app fails with an exception:

```
        static void Main(string[] args)
        {
            var schemaReader = XmlTextReader.Create("main.xsd");
            var thisSchema = XmlSchema.Read(schemaReader, null);
            var thisSchemaSet = new XmlSchemaSet();
            thisSchemaSet.Add(thisSchema);
            thisSchemaSet.Compile();
        }
```

The exception is as follows:

```
System.Xml.Schema.XmlSchemaException occurred
  HResult=0x80131941
  Message=Type 'http://www.xbrl.org/2003/instance:monetaryItemType' is not declared.
  Source=<Cannot evaluate the exception source>
  StackTrace:
   at System.Xml.Schema.XmlSchemaSet.InternalValidationCallback(Object sender, ValidationEventArgs e)
   at System.Xml.Schema.BaseProcessor.SendValidationEvent(XmlSchemaException e, XmlSeverityType severity)
   at System.Xml.Schema.BaseProcessor.SendValidationEvent(XmlSchemaException e)
   at System.Xml.Schema.Compiler.CompileElement(XmlSchemaElement xe)
   at System.Xml.Schema.Compiler.Compile()
   at System.Xml.Schema.Compiler.Execute(XmlSchemaSet schemaSet, SchemaInfo schemaCompiledInfo)
   at System.Xml.Schema.XmlSchemaSet.Compile()
   at SchemaSetTest.Program.Main(String[] args) in c:\users\jefff\documents\visual studio 2017\Projects\SchemaSetTest\SchemaSetTest\Program.cs:line 14
```

The type in question (`http://www.xbrl.org/2003/instance:monetaryItemType`) is available from the imported schema named on the main XSD's `<import>` element. This code works on .NET Framework. Is this a known issue with `XmlSchemaSet` in .NET Core 2.0?

**Comment by sepidehMS** [comment]
@JeffFerguson That's happening because the inner schemas have not been imported. In .NET Core, resolving URIs inside a schema is not allowed by default and that's by design. You will need to add a XmlUrlResolver to your SchemaSet like: `thisSchemaSet.XmlResolver = new XmlSchemaSet();`

**Comment by JeffFerguson** [comment]
I can't assign an `XmlSchemaSet` to an `XmlResolver`, so the statement `thisSchemaSet.XmlResolver = new XmlSchemaSet();` won't compile. I'm assuming that you mean an object of a type like `XmlUrlResolver()`? That code addition still throws an exception:

```
            var thisSchemaSet = new XmlSchemaSet();
            thisSchemaSet.XmlResolver = new XmlUrlResolver();
            thisSchemaSet.Add(thisSchema);
            thisSchemaSet.Compile();
```

**Comment by sepidehMS** [comment]
@JeffFerguson Sorry for the confusion! Yes, I meant `thisSchemaSet.XmlResolver = new XmlUrlResolver();`. Fixed it in the original comment. After adding the XmlUrlResolver, I can't repro this. Could you please tell me which platform you are running on and what `dotnet --info` returns?

**Comment by JeffFerguson** [comment]
Certainly! I'd be happy to.

**dotnet --info output**
```
C:\WINDOWS\system32>dotnet --info
.NET Command Line Tools (2.0.0-preview1-005977)

Product Information:
 Version:            2.0.0-preview1-005977
 Commit SHA-1 hash:  414cab8a0b

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.0.0-preview1-005977\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0-preview1-002111-00
  Build    : 1ff021936263d492539399688f46fd3827169983
```
**main.xsd**
```
<?xml version="1.0" encoding="utf-8"?>
<schema targetNamespace="http://example.com/xbrl/taxonomy" xmlns="http://www.w3.org/2001/XMLSchema" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:xbrli="http://www.xbrl.org/2003/instance" xmlns:link="http://www.xbrl.org/2003/linkbase" xmlns:xlink="http://www.w3.org/1999/xlink">
  <import namespace="http://www.xbrl.org/2003/instance" schemaLocation="http://www.xbrl.org/2003/xbrl-instance-2003-12-31.xsd"/>
  <element name="changeInRetainedEarnings" type="xbrli:monetaryItemType" substitutionGroup="xbrli:item" id="a1" xbrli:periodType="duration"/>
  <element name="fixedAssets" type="xbrli:monetaryItemType" substitutionGroup="xbrli:item" id="a2" xbrli:balance="debit" xbrli:periodType="instant"/>
</schema>
```
**Console app**
```
        static void Main(string[] args)
        {
            var schemaReader = XmlTextReader.Create("main.xsd");
            var thisSchema = XmlSchema.Read(schemaReader, null);
            var thisSchemaSet = new XmlSchemaSet();
            thisSchemaSet.XmlResolver = new XmlUrlResolver();
            thisSchemaSet.Add(thisSchema);
            thisSchemaSet.Compile();
        }
```
**Exception**
```
System.Xml.Schema.XmlSchemaException occurred
  HResult=0x80131941
  Message=Type 'http://www.xbrl.org/2003/instance:monetaryItemType' is not declared.
  Source=<Cannot evaluate the exception source>
  StackTrace:
   at System.Xml.Schema.XmlSchemaSet.InternalValidationCallback(Object sender, ValidationEventArgs e)
   at System.Xml.Schema.BaseProcessor.SendValidationEvent(XmlSchemaException e, XmlSeverityType severity)
   at System.Xml.Schema.BaseProcessor.SendValidationEvent(XmlSchemaException e)
   at System.Xml.Schema.Compiler.CompileElement(XmlSchemaElement xe)
   at System.Xml.Schema.Compiler.Compile()
   at System.Xml.Schema.Compiler.Execute(XmlSchemaSet schemaSet, SchemaInfo schemaCompiledInfo)
   at System.Xml.Schema.XmlSchemaSet.Compile()
   at SchemaSetTest.Program.Main(String[] args) in C:\Users\jefff\documents\visual studio 2017\Projects\SchemaSetTest\SchemaSetTest\Program.cs:line 15
```

**Comment by sepidehMS** [comment]
@JeffFerguson This is related to the bug fixed by https://github.com/dotnet/corefx/pull/19812. Unfortunately, this fix was not included in 2.0.0-preview1-005977 release. In the meantime, you could use our daily builds of .NET Core 2.0 to get the fixed behavior. Here's documentation on how to do that: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md

Closing the bug...
